receiver/httpreceiver/logs.go

```diff
@@ -0,0 +1,256 @@
+	r.wg.Wait()
+	return nil
+
+}
+
+// handleRequest is the function the server uses for requests; calls ConsumeLogs
+func (r *httpLogsReceiver) handleRequest(rw http.ResponseWriter, req *http.Request) {
+	// path was configured && this req.URL does not match it
+	if r.path != "" && req.URL.Path != r.path {
+		rw.WriteHeader(http.StatusNotFound)
+		r.logger.Debug("received request to path that does not match the configured path", zap.String("request path", req.URL.Path))
+		return
+	}
+
+	// read in request body
+	var payload []byte
+	if req.Header.Get("Content-Encoding") == "gzip" {
+		r.logger.Debug("req header has Content-Encoding set to gzip")
+		reader, err := gzip.NewReader(req.Body)
+		if err != nil {
+			rw.WriteHeader(http.StatusUnprocessableEntity)
+			r.logger.Debug("got payload with gzip compression but failed to read", zap.Error(err))
+			return
+		}
+		defer reader.Close()
+
+		// read the decompressed response body
+		payload, err = io.ReadAll(reader)
+		if err != nil {
+			rw.WriteHeader(http.StatusUnprocessableEntity)
+			r.logger.Debug("got payload with gzip compression but failed to read uncompressed payload", zap.Error(err))
+			return
+		}
+	} else {
+		r.logger.Debug("req header does not have Content-Encoding set to gzip")
+		var err error
+		payload, err = io.ReadAll(req.Body)
+		if err != nil {
+			rw.WriteHeader(http.StatusUnprocessableEntity)
+			r.logger.Debug("failed to read logs payload", zap.Error(err), zap.String("remote", req.RemoteAddr))
+			return
+		}
+	}
+
+	// parse []byte into map structure
+	logs, err := parsePayload(payload)
+	if err != nil {
+		rw.WriteHeader(http.StatusUnprocessableEntity)
+		r.logger.Error("failed to convert log request payload to maps", zap.Error(err))
+		return
+	}
+
+	// consume logs after processing
+	if err := r.consumer.ConsumeLogs(req.Context(), r.processLogs(pcommon.NewTimestampFromTime(time.Now()), logs)); err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		r.logger.Error("failed to consume logs", zap.Error(err))
+		return
+	}
+
+	rw.WriteHeader(http.StatusOK)
+}
+
+// processLogs transforms the parsed payload into plog.Logs
+func (r *httpLogsReceiver) processLogs(now pcommon.Timestamp, logs []map[string]interface{}) plog.Logs {
+	pLogs := plog.NewLogs()
+	resourceLogs := pLogs.ResourceLogs().AppendEmpty()
+	scopeLogs := resourceLogs.ScopeLogs().AppendEmpty()
+
+	for _, log := range logs {
+		logRecord := scopeLogs.LogRecords().AppendEmpty()
+
+		logRecord.SetObservedTimestamp(now)
+
+		if err := logRecord.Body().SetEmptyMap().FromRaw(log); err != nil {
+			r.logger.Warn("unable to set log body", zap.Error(err))
+		}
+	}
+
+	return pLogs
+}
+
+// parsePayload transforms the payload into []map[string]interface structure
+func parsePayload(payload []byte) ([]map[string]interface{}, error) {
+	rawLogsArray := []json.RawMessage{}
+	rawLogObject := json.RawMessage{}
+	if err := json.Unmarshal(payload, &rawLogsArray); err != nil {
```

Comment:
I'm curious if it would be more efficient to find the first non-whitespace character and check for `[` or `{` to do specify the decoding since unmarshalling is sort of a heavy call. Maybe not necessary for initial PR but an idea.

Reply:
no that makes sense to me, especially if Unmarshal is a heavier call

Reply:
I was more thinking of something like this, 

```
// parsePayload transforms the payload into []map[string]interface structure
func parsePayload(payload []byte) ([]map[string]interface{}, error) {
	firstChar := seekFirstNonWhitespace(string(payload))
	switch firstChar {
	case "{":
		rawLogObject := json.RawMessage{}
		if err := json.Unmarshal(payload, &rawLogObject); err != nil {
			return nil, err
		}
		return parseJSONObject(payload)
	case "[":
		rawLogsArray := []json.RawMessage{}
		if err := json.Unmarshal(payload, &rawLogsArray); err != nil {
			return nil, err
		}
		return parseJSONArray(rawLogsArray)
	default:
		return nil, errors.New("unsupported payload format, expected either a JSON object or array")
	}
}

func seekFirstNonWhitespace(s string) string {
	firstChar := ""
	for _, r := range s {
		if unicode.IsSpace(r) {
			continue
		}
		firstChar = string(r)
		break
	}
	return firstChar
}
```

You're current solution does not account for whitespace in the payload.

i.e. 
			payload:      `[{"status": "info", "timestamp": "8402734234", "message": {"data": "hello world"}, "hostname": "localhost"}]`,

will work but

			payload:      ` [{"status": "info", "timestamp": "8402734234", "message": {"data": "hello world"}, "hostname": "localhost"}]`,

will not

Reply:
I also think the parseJSONObject function can just return the first unmarshal no? unless i'm missing something

Reply:
yea the whitespace thing is fair, and the switch statement does look cleaner

Reply:
the parseJSONObject func doesn't return the first unmarshal cus the parsePayload func returns an array of the map[string]interface{}, so just creating the array in the parseJSONObject so it fits well with the other functions

Reply:
It seems like we don't even need that func though since we can inline declare a slice:

```go
switch firstChar {
	case "{":
		rawLogObject := map[string]interface{}{}
		if err := json.Unmarshal(payload, &rawLogObject); err != nil {
			return nil, err
		}
		return []map[string]interface{}{rawLogObject}, nil
	case "[":
		rawLogsArray := []json.RawMessage{}
		if err := json.Unmarshal(payload, &rawLogsArray); err != nil {
			return nil, err
		}
		return parseJSONArray(rawLogsArray)
	default:
		return nil, errors.New("unsupported payload format, expected either a JSON object or array")
	}
```